### DISCUSSION TITLE
Are dynamic types possible?

### DISCUSSION THREAD

**Discussion by shortwavedave** [swaggest/rest#169] [Q&A] [answered]
Hello, I would like to parse some remote data (per user request) and add data models/interactors dynamically - is that possible?

**Comment by vearutop** [comment, marked as answer]
Please check a newly added example https://github.com/swaggest/rest/blob/v0.2.58/_examples/advanced/dynamic_schema.go.

**Reply by shortwavedave** [reply]
Thank you so much!

**Comment by shortwavedave** [comment]
@vearutop One more question on this subject: I wanted to set the definition name in the schema, e.g.,

```golang
	dynOut := dynamicOutput{}
	dynOut.Struct.DefName = "Test123"
	dynOut.Struct.Fields = []jsonschema.Field{
		{Name: "Foo", Value: 123, Tag: `header:"foo" enum:"123,456,789"`},
		{Name: "Bar", Value: "abc", Tag: `json:"bar"`},
		{Name: "_", Value: new(struct{}), Tag: `query:"_" additionalProperties:"false" description:"aaMyObj is my object." title: "BLAH"`},
	}
```
But the output always appears to use the structure name rather than "Test123", here's the output:
```json
   "DynamicOutput": {
    "type": "object",
    "properties": {
     "bar": {
      "type": "string"
     },
     "status": {
      "type": "string"
     }
    },
    "additionalProperties": false,
    "description": "aaMyObj is my object."
   },
```
Do you know how i might change the definition name? The use case is that I would like to reuse that dynamic structure to generate many different definitions

**Reply by shortwavedave** [reply]
Additionally, is there a pattern for replacing struct tags with something dynamic? How could we indicate to the generator to set the `json:bar` to a variable rather than literal?

**Reply by vearutop** [reply]
There was an issue in `DefName` handling for embedded `Struct`, please try `go get -u github.com/swaggest/jsonschema-go@v0.3.59`.

> Additionally, is there a pattern for replacing struct tags with something dynamic? How could we indicate to the generator to set the json:bar to a variable rather than literal?

I'm not sure if I understand the case, could you elaborate or make an example?

**Reply by shortwavedave** [reply]
@vearutop Thank you for the support, I will test your update later today. Here's a non-working example of the case I am describing:

```golang
type DynamicStruct struct {
	jsonschema.Struct

	jsonFields   map[string]interface{}
	headerFields map[string]string

	Name string
}

func ExtractStructs(value Node, outerDynStruct *DynamicStruct) []DynamicStruct {
	res := []DynamicStruct{}
	i, _ := Node.Fields()
	for i.Next() {
		switch i.Kind() {
		case Struct:
			// This field is a struct, so create a new dynamic struct and add
			// it to the list of structs
			newDynStruct := DynamicStruct{}
			newDynStruct.Struct.DefName = toCamel(i.Label())
			// Append the structs to the results
			res = append(res, newDynStruct)
			res = append(res, newStructs...)
		default:
			newField := jsonschema.Field{Name: toCamel(i.Label()), Value: "", Tag: fmt.Sprintf("json:%s",i.Label())}
			outerDynStruct.Struct.Fields = append(outerDynStruct.Struct.Fields, newField)
		}

	}
	return res
}

func parse_remote_schema() []usecase.Interactor {
	interactors := []usecase.Interactor{}

	// WalkSchema(value, before)
	dummy := DynamicStruct{}
	res := ExtractStructs(value, &dummy)
	for i, r := range res {
		u := usecase.NewIOI(r, r, func(ctx context.Context, input, output interface{}) (err error) {
			in := input.(DynamicStruct)
			out := output.(*DynamicStruct)
			return nil
		})

		u.SetTitle("Schema DataModel")
		u.SetDescription("Blah blah blah.")
		interactors = append(interactors, u)
	}
	return interactors
}
```

The relevant line is:
```golang
newField := jsonschema.Field{Name: toCamel(i.Label()), Value: "", Tag: fmt.Sprintf("json:%s",i.Label())}
```
This is impossible because struct tags must be literals and cannot be dynamic like this. My current thinking is that I must use one of the Intercept callbacks (I think InterceptProperty) to then read some private data from the dynamic structure and add the tags manually, although I am not certain how to accomplish this yet. Aside from the json label, I would also like to set the constrains and query options, etc. 

If you would like a more complete example, please let me know.

**Reply by vearutop** [reply]
The type of `Tag` is `reflect.StructTag`, and also the name needs to be in quotes. Please check if this code works for you.
```go
newField := jsonschema.Field{Name: toCamel(i.Label()), Value: "", Tag: reflect.StructTag(fmt.Sprintf(`json:"%s"`,i.Label()))}
```

**Reply by shortwavedave** [reply]
@vearutop This all appears to be working. Thank you again for the support, I really appreciate it
